src/app/modules/spcp/sp.oidc.client.cache.ts

```diff
@@ -0,0 +1,211 @@
+import axios from 'axios'
+import { createPublicKey } from 'crypto'
+import jwkToPem from 'jwk-to-pem'
+import NodeCache from 'node-cache'
+import { BaseClient, Issuer } from 'openid-client'
+import { timeout } from 'promise-timeout'
+
+import { JwkError } from './sp.oidc.client.errors'
+import {
+  CryptoKeys,
+  PublicJwks,
+  Refresh,
+  SecretJwks,
+  SpOidcClientCacheConstructorParams,
+} from './sp.oidc.client.types'
+import {
+  isEC,
+  retryPromiseForever,
+  retryPromiseThreeAttempts,
+} from './sp.oidc.util'
+
+// Name of keys in cache
+const BASE_CLIENT_NAME = 'baseClient'
+const NDI_PUBLIC_KEY_NAME = 'ndiPublicKeys'
+const EXPIRY_NAME = 'expiry'
+
+/**
+ * Cache class which provides read-through capability and refresh-ahead before expiry
+ * Handles discovery and retrieval of NDI's public jwks
+ */
+export class SpOidcClientCache {
+  /**
+   * Cache to store NDI's keys and client
+   * @private
+   * accessible only for testing
+   */
+  _cache: NodeCache
+
+  /**
+   * Stores the refresh promise so that there is at most
+   * one in-flight refresh at any point in time.
+   * Stored promise is always pending
+   * @private
+   * accessible only for testing
+   */
+  _refreshPromise?: Promise<Refresh>
+
+  #spOidcNdiDiscoveryEndpoint: string
+  #spOidcNdiJwksEndpoint: string
+  #spOidcRpClientId: string
+  #spOidcRpRedirectUrl: string
+  #spOidcRpSecretJwks: SecretJwks
+
+  /**
+   * Constructor for cache
+   * On instantiation, trigger a refresh to populate the cache
+   */
+  constructor({
+    options,
+    spOidcNdiDiscoveryEndpoint,
+    spOidcNdiJwksEndpoint,
+    spOidcRpClientId,
+    spOidcRpRedirectUrl,
+    spOidcRpSecretJwks,
+  }: SpOidcClientCacheConstructorParams) {
+    this.#spOidcNdiDiscoveryEndpoint = spOidcNdiDiscoveryEndpoint
+    this.#spOidcNdiJwksEndpoint = spOidcNdiJwksEndpoint
+    this.#spOidcRpClientId = spOidcRpClientId
+    this.#spOidcRpRedirectUrl = spOidcRpRedirectUrl
+    this.#spOidcRpSecretJwks = spOidcRpSecretJwks
+    this._cache = new NodeCache(options)
+
+    // On expiry, refresh cache.
+    this._cache.on('expired', () => this.refresh())
+
+    // Trigger refresh on instantiation to populate cache.
+    void this.refresh()
```

Comment:
Even though `refresh()` is a never-failing promise (infinite retry), continuous failures on startup are problematic because the Singpass functionality would be broken. Is it possible to have some custom logging for the startup failures anyway?

Hmm, or perhaps we do have enough information after all, since we will have the rejection logs, AND we should see independently the SingPass login errors, correct? 🤔

Reply:
Thanks! I think no harm to make additional logs to facilitate troubleshooting, I've included it

Reply:
Thanks Shuli!

The current login implementation might be a little noisy (one or 2 per singpass login attempts, rather than one per refresh attempt), but I think that's OK for now 👍